### PR TITLE
UnitControl: Improve interactions + add unit parsing

### PR DESCRIPTION
**Pull request by ItsJonQ**
## Description

Screen Capture on 2020-05-13 at 13-21-43

(GIF: Demo of dragging UnitControl to update Cover height)

This update adds rich features and interactions to the UnitControl component. Originally, this work was in support of adding [padding controls to the Cover block](https://github.com/WordPress/gutenberg/pull/21492). This update abstracts just the UnitControls from that effort.

Notable features for UnitControl include:

* Drag (mouse) to update values
* Unit parsing from input. Example: `10px` is parsed and understood as `[10, 'px']`
* Improved step increment calculations (holding <kbd>Shift</kbd> when updating)

UnitControl and NumberControl are now powered by a new lower level primitive: InputControl.

InputControl handles the majority of the interaction and event logic. Components using it, like NumberControl, are able to tap into the state flow and add features like "drag to update".

Notable features for InputControl include:

* Floating label UI interaction - Not enabled by default
* Styled with G2 designs, out of the box

## How has this been tested?

* Tested in Storybook
* Tested in local Gutenberg
* Tested in Jest (via unit tests)

### Storybook

* Run `npm install` to download dependencies
* Fire up storybook with `npm run storybook:dev`
* Search for `unit control`
* Play around with the component to ensure it works as expected

### Gutenberg

* Run `npm install` to download dependencies
* Fire up local development with `npm run dev`
* Add a Cover block
* Adjust the height. Ensure it works as expected (Note: Unit parsing feature is not enabled. This would involve updating the Cover block component a bit)

## Types of changes

* Adding experimental components

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/input-control/README.md
Comment (ZebulanStanphill) [inline packages/components/src/input-control/README.md:45]: This description seems a bit vague; even I'm not entirely sure what it means. I assume it means the label appears inside the input until the input is focused? Also, is this prop even a good idea, accessibility-wise?

Also, all this type info belongs in a JSDoc comment, doesn't it?

File: packages/components/src/input-control/index.js
Comment (ZebulanStanphill) [inline packages/components/src/input-control/index.js:4]: I've seen this code pattern used in several places, but given that we can now use optional chaining, I think that instead of doing this, we could just let the functions be `undefined` by default. I'm just guessing (and could be completely wrong about this), but it seems like that would perform slightly better. To avoid calling functions that are `undefined`, we would use optional chaining like so:

```js
onBlur?.( event )`
```

Again, this is just an idea; it probably doesn't matter much either way.

File: packages/components/src/input-control/state.js
Comment (mcsf) [inline packages/components/src/input-control/state.js:62]: Looking at this with a little more distance, it's worth pointing out that this is very similar to a function composition, to the point that maybe it ought to be named `composeReducers`. In the traditional `combineReducers`, the order doesn't matter, as each subreducer gets its own branch in the resulting reducer. In contrast, here order does matter in that a subreducer can squash another's state.

File: packages/components/src/input-control/state.js
Comment (mcsf) [inline packages/components/src/input-control/state.js:64]: Should we keep this flattening? Do you have concrete examples of when you'd want to traverse a deep list of reducers? Otherwise we might keep a stricter interface and only take `...fns` with `fns :: Array<Reducer>`.

I think being strict about this idea that reducers are _linear input_ rather than tree-like is also more consistent with my above comment on `composeReducers`. Saves us and API consumers from confusion. :)

**Review by mcsf — COMMENTED** [review]
There's a lot to unpack here and I've barely scratched the surface, I feel! That said, in terms of UX this seems pretty solid, and I think there are some good overall patterns in the implementation.

File: packages/components/src/input-control/index.js
Comment (ItsJonQ) [reply packages/components/src/input-control/index.js:4]: Interesting idea! I thought about it for a bit (considering optional chaining is a new thing)

From my experience in the React library/module space, I've found it smoother for callbacks to fire without guards. It kept the code tidier and there was less to think about.

That being said, optional chaining does tidy it up a lot.

I'm perhaps biased in that I still prefer `noop` callbacks.
I'm open to using optional chaining instead, if that's what others prefer 👍

File: packages/components/src/unit-control/utils.js
Comment (ItsJonQ) [reply packages/components/src/unit-control/utils.js:63]: I didn't know about `keyBy`! Just curious, what is the advantage between the approaches? (Asking to learn. Totally open to changing it 😊 )

File: packages/components/src/unit-control/utils.js
Comment (ItsJonQ) [reply packages/components/src/unit-control/utils.js:70]: Thank you! Will update

File: packages/components/src/unit-control/utils.js
Comment (ItsJonQ) [reply packages/components/src/unit-control/utils.js:54]: I just tested it out! It turns out `.trim()` is needed as `value` is used later for unit parsing.

File: packages/components/src/input-control/README.md
Comment (ItsJonQ) [reply packages/components/src/input-control/README.md:7]: Oops! Forgot to remove this from copy/paste 🙈 

Thank you!

File: packages/components/src/input-control/state.js
Comment (ItsJonQ) [reply packages/components/src/input-control/state.js:64]: Ah, good catch! Yes, since we're explicitly using it between the layers (example below), we no longer need to flatten the reducer args.

```jsx
<Input
	__unstableStateReducer={ combineStateReducers(
		numberControlStateReducer,
		stateReducer
	) }
/>
```

File: packages/components/src/input-control/state.js
Comment (ItsJonQ) [reply packages/components/src/input-control/state.js:62]: Oo! Interesting observation. I like this a lot! Will make that change

File: packages/components/src/input-control/README.md
Comment (ItsJonQ) [reply packages/components/src/input-control/README.md:43]: Thanks for noticing this! I'm not sure how best to describe this interaction/design.

It works like the TextFields from Material UI:
https://material-ui.com/components/text-fields/

From a design perspective, it offers more versatility, especially if controls need to be combined in cramped spaces.

As for as accessibility goes, how its setup, I believe it it acceptable from both a motion perspective and aria label + HTML markup perspective.

cc'ing some a11y folks!

@enriquesanchez / @diegohaz  Are there any concerns for having floating labels in Inputs like these?

Screen Shot 2020-05-14 at 3 07 28 PM


https://material-ui.com/components/text-fields/

Thank you!

File: packages/components/src/unit-control/utils.js
Comment (ZebulanStanphill) [inline packages/components/src/unit-control/utils.js:60]: Shouldn't this be:
```suggestion
	let unit = unitMatch !== undefined ? unitMatch : '';
```

File: packages/components/src/input-control/backdrop.js
Comment (ZebulanStanphill) [inline packages/components/src/input-control/backdrop.js:10]: Components should have [JSDoc comments with type info](https://developer.wordpress.org/block-editor/contributors/develop/coding-guidelines/#documenting-wordpresselement-react-components).

File: packages/components/src/input-control/backdrop.js
Comment (ZebulanStanphill) [inline packages/components/src/input-control/backdrop.js:13]: What's the difference between these two?

File: packages/components/src/input-control/backdrop.js
Comment (ZebulanStanphill) [inline packages/components/src/input-control/backdrop.js:14]: Should this have a default argument?

File: packages/components/src/input-control/backdrop.js
Comment (ItsJonQ) [reply packages/components/src/input-control/backdrop.js:13]: One tracks the setting preference (and validity) of it via the `isFloatingLabel` prop. The other tracks the floating state.

The varying combinations adjust the styles for the HTML elements

File: packages/components/src/input-control/backdrop.js
Comment (ItsJonQ) [reply packages/components/src/input-control/backdrop.js:10]: Just to confirm, it sounds like we need to add this to every component, correct?

Looking at previous components, there doesn't appear to be many examples of JSDoc comments:

https://github.com/WordPress/gutenberg/blob/master/packages/components/src/select-control/index.js

I'm guessing the goal is to add these over time to increase doc coverage?

File: packages/components/src/input-control/backdrop.js
Comment (ItsJonQ) [reply packages/components/src/input-control/backdrop.js:14]: Ah yes. Thank you for catching

File: packages/components/src/input-control/backdrop.js
Comment (ZebulanStanphill) [reply packages/components/src/input-control/backdrop.js:13]: Perhaps `isFloatingLabel` should be renamed to `useFloatingLabel` or `floatingLabel` or something like that?

Comment (ItsJonQ) [reply packages/components/src/input-control/backdrop.js:13]: I don't think we can use the `use` prefix, since that's associated with hooks. I think ESLint might actually complain.

Prefixing with `is` typically indicates a `boolean` value, which this one is.

Comment (ZebulanStanphill) [reply packages/components/src/input-control/backdrop.js:13]: How about `isFloatingLabelStyle` or `isFloatingLabelType`?

Comment (ItsJonQ) [reply packages/components/src/input-control/backdrop.js:13]: Hmm.. I'm finding those names trickier to understand. I'm not sure what the best names may be.

At the moment, both props are shared throughout various sub-components of `InputControl` for both functionality and aesthetics. Because of that, they are consistent within `InputControl`. Unless the current names are overly confusing or broken, I think it may be simpler to keep them.

File: packages/components/src/input-control/README.md
Comment (diegohaz) [reply packages/components/src/input-control/README.md:43]: I don't have strong opinions about that. But there's this article with arguments against floating labels: [Floating labels are problematic](https://medium.com/simple-human/floating-labels-are-a-bad-idea-82edb64220f6)

And this other one is a response to that: [Are float labels really that problematic after all?](http://mds.is/still-floating-labels/)

Comment (enriquesanchez) [reply packages/components/src/input-control/README.md:43]: I always check Adam Silver's articles when in doubt about forms and a11y. That first link Diego shared is from him :)

- Sorry if this is obvious, but where is this floating label going to be used? Only on this control or on all input fields?
- I don't see the floating label in the gif example above, but how would it look like when the floating label is longer than the field itself?

Comment (ItsJonQ) [reply packages/components/src/input-control/README.md:43]: @diegohaz + @enriquesanchez Thanks for replying!

> where is this floating label going to be used? Only on this control or on all input fields?

It's situational! By default, floating labels are **disabled**. They're useful if the input is narrower or used in compact spaces. For example, within a smaller Popover.

> but how would it look like when the floating label is longer than the field itself?

Ah! the GIF above does not have a floating label. The best example of this interaction would be from the Material TextField demo:

https://material-ui.com/components/text-fields/

My screen capture -> GIF software doesn't do a good job at capturing the various stages.

The key frames work like this ✨ 

No value, label works like a placeholder:

Screen Shot 2020-05-19 at 12 45 31 PM

No value, focused, the label floats up:

Screen Shot 2020-05-19 at 12 46 20 PM

With value, the label stays up:
Screen Shot 2020-05-19 at 12 46 27 PM

Comment (ZebulanStanphill) [reply packages/components/src/input-control/README.md:43]: That doesn't really answer the question, though: what happens when the field label is longer than the field itself?

Comment (ItsJonQ) [reply packages/components/src/input-control/README.md:43]: Ah! Gotcha. I misread. My apologies. There isn't handling of this at the moment.

The Material TextField doesn't handle this use case too well either:
https://codesandbox.io/s/material-demo-e49dz

I can try some truncation techniques really quickly. I don't think this should be a blocker though.

Comment (ZebulanStanphill) [reply packages/components/src/input-control/README.md:43]: The big issue is that input labels have to be translated, and some languages will use much longer words than others. An input with a floating label of "left" might be correctly sized, but once you translate that to another language, it may suddenly be twice as long.

WordPress generally prefers decisions over options. If there's not a really good reason to provide an alternate input style, we shouldn't provide it. I think it's okay to merge this PR since the `InputControl` will be experimental, but I don't think we should change the component to stable until we know whether the floating label options should be kept or not.

Comment (ItsJonQ) [reply packages/components/src/input-control/README.md:43]: Screen Capture on 2020-05-19 at 13-52-02

This is the best I could do given the (surprisingly) limited CSS rendering control over `<fieldset>`.

> I don't think we should change the component to stable until we know whether the floating label options should be kept or not

Yup! I agree. We can also drop the floating label feature if it's causing issues.

Comment (enriquesanchez) [reply packages/components/src/input-control/README.md:43]: Thanks @ItsJonQ! I was trying to make a point with my question, apologies if it wasn't clear 😃.

I feel that at this point there's some uncertainty re: floating labels (a11y, localization, placement, etc.) that my recommendation is to stick to regular ol' fashioned text labels.

File: packages/components/src/unit-control/utils.js
Comment (mcsf) [reply packages/components/src/unit-control/utils.js:64]: Ah, well, the idea was that you'd have a plain JS object as returned by `keyBy`:

```js
keyBy(
  [
    { value: 'px', label: 'px', default: 0 },
    { value: '%', label: '%', default: 10 },
  ],
  'value'
)
// returns
{
  px: { value: 'px', label: 'px', default: 0 },
  '%': { value: '%', label: '%', default: 10 }
}
```

which would be handy when directly accessing properties: `obj[ value ]`, in contrast with `arr.find( … )`.

Again, it's not a blocker, but something to consider depending on how frequently this will run!

File: packages/components/src/input-control/backdrop.js
Comment (mcsf) [reply packages/components/src/input-control/backdrop.js:10]: When dealing with _private_ components such as Backdrop, for now, I wouldn't let this block development.

File: packages/components/src/input-control/state.js
Comment (mcsf) [inline packages/components/src/input-control/state.js:61]: Consider optimising this by making sure that `fn( ...args )` actually returns a non-empty object before concatenating with `state` in a new object. Otherwise, we can avoid returning new references. Something like:

```js
const fnState = fn( ...args );
return isEmpty( fnState ) ? state : { ...state, ...fnState };
```

Whether you'd like to tackle this now or as a separate task, it's your call. :) It may also depend on actual usage of `composeStateReducers`.

File: packages/components/src/input-control/state.js
Comment (mcsf) [inline packages/components/src/input-control/state.js:174]: Seems like this is no longer needed, there's nothing to compose.

File: packages/components/src/input-control/state.js
Comment (mcsf) [inline packages/components/src/input-control/state.js:52]: Minor point: in this function and the rest of the file I'd replace all mentions of "combine", as they may become ambiguous. In some cases, "compose" is apt; in others, I think we can drop the adjectives altogether, e.g. `const reducer = composeStateReducers( fooReducer, barReducer )`.

**Review by mcsf — APPROVED** [review]

File: packages/components/src/input-control/README.md
Comment (ItsJonQ) [reply packages/components/src/input-control/README.md:43]: > I feel that at this point there's some uncertainty re: floating labels (a11y, localization, placement, etc.) that my recommendation is to stick to regular ol' fashioned text labels.

I agree! Ol' fashion text labels are **default** 😊 

The floating labels are a new feature, which can be experimentally trialed in certain scenarios

File: packages/components/src/unit-control/utils.js
Comment (ZebulanStanphill) [reply packages/components/src/unit-control/utils.js:64]: If it does turn out to be a performance concern, perhaps we ought to be keying them from the start when the units are added.

File: packages/components/src/input-control/state.js
Comment (ItsJonQ) [reply packages/components/src/input-control/state.js]: Ah! Good catch 🤗 . Will push a fix for this

File: packages/components/src/input-control/state.js
Comment (ItsJonQ) [reply packages/components/src/input-control/state.js]: I think this makes sense! Thank you for your suggestion! I just tested, and it seems to work as expected 👍

File: packages/components/src/unit-control/utils.js
Comment (ItsJonQ) [reply packages/components/src/unit-control/utils.js]: @mcsf Thank you for the context! It doesn't run super often (not on every keystroke). Still, I think it's a worth while perf boost!

**Comment by ItsJonQ** [comment]
Thanks for the feedback all! I'll merge this once Travis is green and happy ✨

File: packages/components/src/unit-control/utils.js
Comment (mcsf) [inline packages/components/src/unit-control/utils.js:64]: @ItsJonQ: Ah, I wasn't clear, but I meant that a keyed version of `units` could be kept where it can be reused — so possibly as a module constant. Otherwise we're still iterating over the collection every time!

Comment (ItsJonQ) [reply packages/components/src/unit-control/utils.js:64]: @mcsf Ah, gotcha! It's a little tricky, as the units coming in may not be constant. Ultimately, this is determined by the `add_theme_support` config within the context of Gutenberg.

We could memoize it? I'm wondering if it's worth it, considering this operation is quite small (within the grand scheme of things)

Comment (mcsf) [reply packages/components/src/unit-control/utils.js:64]: We could memoise, yeah, but I don't want to take more of your time on this now, so I suggest reverting to the initial array and `find` situation, and we can improve from there. :)

Comment (ItsJonQ) [reply packages/components/src/unit-control/utils.js:64]: Sounds good! Thank you @mcsf !!

**Comment by ItsJonQ** [comment]
Rebasing with latest `master`

**Comment by ItsJonQ** [comment]
Accidentally clicked "Close and comment" 🙈

**Comment by ItsJonQ** [comment]
Hmm... looks like Travis has been failing the E2E (not just this PR, but for others as well)

**Comment by ItsJonQ** [comment]
Yay! Travis is happy! Merging :D

**Comment by ItsJonQ** [comment]
Thanks again everyone! I can't wait to use these 🎉

**Comment by afercia** [comment]
Late to the party. I came across this while testing the new Navigation page.

> This is the best I could do given the (surprisingly) limited CSS rendering control over `<fieldset>`.

I'm not sure I understand why a fieldset + legend elements are used for this pattern. They're semantic elements meant to group logically-related form controls. The group name is given by the legend element. Instead, in this case:
- both the fieldset and the legend elements have an `aria-hidden=true` attribute so they're hidden from assistive technologies, thus defeating the elements purpose to be _semantic elements_
- in any case, the fieldset does not group any form element: it's an empty fieldset
- to my understanding, the fieldset is absolutely positioned over the form element in question and uses `pointer-events: none;` to allow click-throughs
- at this point why use a fieldset element in the first place? It could be a simple div element
- if there's no `<label>` element correctly associated to the input, or an `aria-label`, the input will be unlabelled even if there's some "visible text": all form controls need to be properly labelled
- lastly, (noticed while testing the new Navigation page), it seems the fieldset is _always rendered_ even when there's no need for it, see the screenshot below:

Screenshot 2020-09-12 at 13 53 35

This is confusing: why we should render a fieldset when it's not needed?

@ItsJonQ can you please shed some light on the reason why a fieldset + legend elements are used here?

Regarding the "floating labels" pattern: in general, anything that "moves" on the screen without user activation is potentially confusing and triggers cognitive load. I'm not sure this pattern should be used without an in-depth discussion between the accessibility and design teams.

In any case, there are better ways to implement "floating labels" that make sure the input is labelled. See for example what GMail does:

floating label gmail visible text and aria label 02

where it's of fundamental importance that the visible text matches the input aria-label.

**Comment by ItsJonQ** [comment]
> This is confusing: why we should render a fieldset when it's not needed?

@afercia Hi there! Thank you for digging in!

I had used `fieldset` as it enabled the borders to render correctly when "floating" the label. I had followed the structure of other examples to achieve this affect. However, as you had noted, there appears to be a better methods (I need to poke at Google/Gmail to figure it out). The current implementation is not worth the degradation of a11y for inputs.

I'll follow up with improvements on this as soon as I can. Would it be okay if I `@mention` you on the PR (when it's available)?

**Comment by afercia** [comment]
> Would it be okay if I @mention you on the PR (when it's available)?

Sure. I'd consider to also post a quick heads up on the Slack accessibility channel so that if I'm not available, someone else can provide feedback. Thanks! ❤️

**Comment by ItsJonQ** [comment]
@afercia Will do! Will also be sure to tag the PR with "Needs Accessibility Feedback"

**Comment by afercia** [comment]
I just noticed that now that the styles from `input-control-styles` are applied to some `<label>` elements, weird things happen as in: clicking the label doesn't focus the associated form element. Will create a separate issue...

**Comment by ItsJonQ** [comment]
> I just noticed that now that the styles from input-control-styles are applied to some <label> elements, weird things happen as in: clicking the label doesn't focus the associated form element. Will create a separate issue...

@afercia I noticed this too when working on the fix. This has been resolved! (Good timing, I'm writing the PR as we speak!)

**Comment by afercia** [comment]
Ah good! I was just writing the new issue :D  Too many `pointer-events: none;`, right? Will wait then and check your PR.

**Pull request merged**
